gapis/perfetto/android/trace.go

```diff
@@ -95,10 +117,11 @@ func Start(ctx context.Context, d adb.Device, a *android.ActivityAction, opts *s
 		}.Bind(ctx)
 		// Before we start the activity, attempt to setup environment if gpu.renderstages is selected.
 		var err error
-		cleanup, err = setupRenderStagesEnvironment(ctx, d, a.Package.Name, opts.PerfettoConfig)
+		c, err := setupRenderStagesEnvironment(ctx, d, a.Package.Name, opts.PerfettoConfig, abi, layers)
 		if err != nil {
 			return nil, cleanup.Invoke(ctx), err
 		}
+		cleanup = cleanup.Then(c)
```

Comment:
Oops, the change to the cleanup actually made things worse. `cleanup` is nil throughout this block. Functions that return a cleanup and an error should only return a cleanup that is expected to be invoked in both success and error situations. Thus to fix your change you'd have to call `c.Invoke(ctx)`, but why create another local var?

Reply:
Fixed